### PR TITLE
(PUP-3263) Adds `architecture` parameter

### PR DESCRIPTION
**Pull request by petems**
Can specify arch wanted when installing a package.

This is extremely rough, needs some help getting it across the line :smile:

### CODE REVIEW

File: lib/puppet/type/package.rb [hunk 1]
```diff
@@ -66,6 +66,7 @@ module Puppet
       :methods => [:package_settings_insync?, :package_settings, :package_settings=]
     feature :virtual_packages, "The provider accepts virtual package names for install and uninstall."
 
+
```
Comment (joshcooper) [inline lib/puppet/type/package.rb]: extra newline

File: lib/puppet/provider/package/yum.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@
   These options should be specified as a string (e.g. '--flag'), a hash (e.g. {'--flag' => 'value'}),
   or an array where each element is either a string or a hash."
 
-  has_feature :install_options, :versionable, :virtual_packages
+  has_feature :install_options, :versionable, :virtual_packages, :package_arch
```
Comment (joshcooper) [inline lib/puppet/provider/package/yum.rb]: Often provider features are "manages_something", perhaps "manages_architecture"? /cc @nfagerlund

File: lib/puppet/type/package.rb [hunk 3]
```diff
@@ -384,6 +385,10 @@ def change_to_s(currentvalue, newvalue)
       desc "A read-only parameter set by the package."
     end
 
+    newparam(:package_arch) do
+      desc "Choose the architecture version of a package"
+    end
+
```
Comment (joshcooper) [inline lib/puppet/type/package.rb]: since this is the package type, can we shorten this to `arch` or `architecture`?

File: lib/puppet/type/package.rb [hunk 4]
```diff
@@ -384,6 +385,10 @@ def change_to_s(currentvalue, newvalue)
       desc "A read-only parameter set by the package."
     end
 
+    newparam(:package_arch) do
```
Comment (joshcooper) [inline lib/puppet/type/package.rb]: Since this is a parameter, puppet won't compare the current arch against the desired arch. So if you had `noarch` installed, but wanted `x86_64` puppet would still think it's in-sync. Would that ever happen in practice? If so, we'd need to make this a property, provided we can always retrieve the current `arch` and compare against the desired value.

**Comment by joshcooper** [comment]
@petems Thanks for submitting this, it would be useful for a number of package providers. rpm and yum already parse the currently installed arch, so it's part of the way there. Also it'd be nice to be able to install platform specific gems. I've rekicked the failing gems as I think we had some CI fixed after you submitted this.

File: lib/puppet/provider/package/yum.rb [hunk 5]
```diff
@@ -151,6 +151,10 @@ def install
       end
     end
 
+    if @resource[:package_arch]
+      wanted = "#{wanted}.#{@resource[:package_arch]}"
+    end
+
```
Comment (joshcooper) [inline lib/puppet/provider/package/yum.rb]: This code path is during install, should we be able to uninstall an arch specific package too?

File: lib/puppet/provider/package/yum.rb (hunk 2 above)
Comment (nfagerlund) [inline lib/puppet/provider/package/yum.rb]: None of the package ones currently follow that standard — see http://docs.puppetlabs.com/references/latest/type.html#package-provider-features for the list. Since the standard doesn't seem to be universal, I don't particularly care, and would default to using the local style for that resource type. (In this case, using the name of the parameter as the name of the feature.)

File: lib/puppet/type/package.rb (hunk 3 above)
Comment (petems) [inline lib/puppet/type/package.rb]: I think architecture, to avoid confusion with the Arch OS. Will fix :+1:

File: lib/puppet/type/package.rb (hunk 4 above)
Comment (petems) [inline lib/puppet/type/package.rb]: Hmm, it's a possibility, I'll change to a property. I'll fix that.

File: lib/puppet/provider/package/yum.rb (hunk 2 above)
Comment (petems) [inline lib/puppet/provider/package/yum.rb]: Changed.

File: lib/puppet/type/package.rb (hunk 1 above)
Comment (petems) [inline lib/puppet/type/package.rb]: Fixed.

File: lib/puppet/provider/package/yum.rb (hunk 5 above)
Comment (petems) [inline lib/puppet/provider/package/yum.rb]: Hmm, possibly. The Yum provider uses the RPM uninstall code, might need some help plumbing that.

**Comment by hkenney** [comment]
@joshcooper This might be ready for merge now that tests are green? Could you take a final look at it?

File: lib/puppet/provider/package/yum.rb (hunk 5 above)
Comment (MikaelSmith) [inline lib/puppet/provider/package/yum.rb]: Did your update attempt to address uninstall? We also wonder about how doing install with a different package architecture would work if the package is already installed. The in_sync check will confirm the package is already installed, requiring you to explicitly uninstall it in one run, then install the other version in another.

**Comment by joshcooper** [comment]
@petems the current implementation makes it possible to specify the architecture during installation, but what about uninstall, e.g. `ensure => absent`? Also, what happens if you have a `noarch` package currently installed, and you specify `architecture => 'x86_64`'. I would expect puppet to install the more specific architecture package, but I don't think the PR supports this.

**Comment by petems** [comment]
@joshcooper Good point, I feel like the arch parameter might need to get bubbled up to the RPM type, and have a method for collecting the current arch value. This would also help with the usecase of uninstall that @MikaelSmith mentioned...

I might try to break this down into a smaller PR just to add the arch parameter as a new property in the RPM T/P, and get that merged first, as the other bits are going to be a bit more complicated.

**Comment by HAIL9000** [comment]
@petems Do you want to continue working off this pull request? Or did you want to close it and open a smaller pull request as you mentioned.

**Comment by kylog** [comment]
@petems have you had a chance to look into this one any more?

**Comment by branan** [comment]
@petems If we don't hear back from you in another week or so we're gonna go ahead and close this for inactivity. Feel free to re-open if you have time to look at it again, or bother us through the normal engineering process to address the ticket.

**Comment by petems** [comment]
@branan Hi, sorry was on holiday, but to be honest I think I'll need engineering help with this. I'll close for now, pending some help.
